File: scratch/diagnose_chars.py
import sys

with open(r'c:\Users\SAAQIB\thirukural\thirukkural.json', 'r', encoding='utf-8') as f:
    content = f.read()

target = "தடை"
import unicodedata
target_nfc = unicodedata.normalize('NFC', target)
target_nfd = unicodedata.normalize('NFD', target)

print(f"Target NFC: {[hex(ord(c)) for c in target_nfc]}")
print(f"Target NFD: {[hex(ord(c)) for c in target_nfd]}")

found_nfc = target_nfc in content
found_nfd = target_nfd in content

print(f"Found NFC: {found_nfc}")
print(f"Found NFD: {found_nfd}")

# Find occurrences and show surrounding characters
if found_nfc:
    idx = content.find(target_nfc)
    print(f"Occurrence at {idx}: {content[idx-10:idx+20]!r}")

File: scratch/diagnose_chars.js
import fs from 'fs';
import path from 'path';

const filePath = path.join('c:', 'Users', 'SAAQIB', 'thirukural', 'thirukkural.json');
const content = fs.readFileSync(filePath, 'utf8');

const target = "தடை";
const target_nfc = target.normalize('NFC');
const target_nfd = target.normalize('NFD');

console.log(`Target NFC Bytes: ${Array.from(target_nfc).map(c => c.charCodeAt(0).toString(16))}`);
console.log(`Target NFD Bytes: ${Array.from(target_nfd).map(c => c.charCodeAt(0).toString(16))}`);

const found_nfc = content.includes(target_nfc);
const found_nfd = content.includes(target_nfd);

console.log(`Found NFC: ${found_nfc}`);
console.log(`Found NFD: ${found_nfd}`);

if (found_nfc) {
    const idx = content.indexOf(target_nfc);
    console.log(`Occurrence at ${idx}: ${content.substring(idx - 10, idx + 20)}`);
}
